File: objectiveai-js/src/functions/executions/request/strategy.ts
import z from "zod";

export const StrategyDefaultSchema = z
  .object({
    type: z.literal("default"),
  })
  .describe("Default strategy for function execution.");
export type StrategyDefault = z.infer<typeof StrategyDefaultSchema>;

export const StrategySwissSystemSchema = z
  .object({
    type: z.literal("swiss_system"),
    pool: z
      .number()
      .int()
      .positive()
      .optional()
      .nullable()
      .describe("How many vector responses for each execution. Default is 10."),
    rounds: z
      .number()
      .int()
      .positive()
      .optional()
      .nullable()
      .describe(
        "How many sequential rounds of comparison. Default is 3."
      ),
  })
  .describe("Swiss system strategy for vector function execution.");
export type StrategySwissSystem = z.infer<typeof StrategySwissSystemSchema>;

export const StrategySchema = z
  .discriminatedUnion("type", [StrategyDefaultSchema, StrategySwissSystemSchema])
  .describe("Strategy for function execution.");
export type Strategy = z.infer<typeof StrategySchema>;

File: objectiveai-sdk-rs/src/functions/executions/request/strategy.rs
use serde::{Deserialize, Serialize};
use schemars::JsonSchema;

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
#[serde(tag = "type", rename_all = "snake_case")]
#[schemars(rename = "functions.executions.request.Strategy")]
pub enum Strategy {
    /// Scalar or Vector
    #[schemars(title = "Default")]
    Default,
    /// Vector
    #[schemars(title = "SwissSystem")]
    SwissSystem {
        /// How many vector responses for each execution
        pool: Option<usize>, // default is 10
        /// How many sequential rounds of comparison
        rounds: Option<usize>, // default is 3
    },
}
